### ISSUE TITLE
Setting python virtual environment

### ISSUE THREAD

**Issue by gjakab4321** [ray-x/navigator.lua#247]
Previously before using the navigator plugin I used this code to set the python virtual environment on a project basis, since I'm jumping between a lot of them. Is it possible to keep the lsp functionality provided by your plugin while extending it with this? I'm trying to set this up for pyright but I'm a bit unsure where and how should I hook this code up.
```python
local util = require('lspconfig/util')
local path = util.path

local function get_python_path(workspace)
  -- Use activated virtualenv.
  if vim.env.VIRTUAL_ENV then
    print(string.format('found virtual env: %s', vim.env.VIRTUALE_ENV))
    return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
  end

  -- Find and use virtualenv in workspace directory.
  for _, pattern in ipairs({ '*', '.*' }) do
    local match = vim.fn.glob(path.join(workspace, pattern, 'poetry.lock'))
    if match ~= '' then
      local venv_base_folder = vim.fn.trim(vim.fn.system('poetry env info -p'))
      print(string.format('venv base folder: %s', venv_base_folder))
      return path.join(venv_base_folder, 'bin', 'python')
    end
  end

  -- Fallback to system Python.
  print('fallback to system python path')
  return exepath('python3') or exepath('python') or 'python'
end

require('lspconfig').pyright.setup({
  on_init = function(client)
    local python_path = get_python_path(client.config.root_dir)
    print(string.format('current python path: %s', python_path))
    client.config.settings.python.pythonPath = python_path
  end,
})

```

**Comment by ray-x** [comment]
In fact you can inject your pyright config with navigator setup
{
lsp = {
  pyright = {
    on_init=function() {..}
}
}
}

but I also pushed a change to allow the above setup to default.

**Comment by gjakab4321** [comment]
Great addition and thank you for your fast response! :slightly_smiling_face:

**Comment by gjakab4321** [comment]
Sorry for reopening the issue but I'm experiencing a new behaviour that affects what I want to achieve (described in the issue description).

So it seems like that previously a new pyright lsp client was started and attached to every relevant buffer which resulted in the running of the `on_init` function every time I opened a python project. This way the virtual environment got set correctly. Currently it seems like a single client gets attached to multiple buffers meaning the on_init functions only runs a single time for the first python project that open.
Can this behaviour be changed?

So to summarize:
- what I want: set the python virtual environment correctly for every project that I open
- what is happening: python virtual environment only gets set for the first project that is opened

**Comment by gjakab4321** [comment]
I'm a bit unsure why but the problem from my previous comments seems to be resolved. I regularly update my plugins so it might have been because of an `nvim-lspconfig` upgrade since navigator uses that too.

Unfortunately, another issue presented itself which is whenever I open a python file 3-4 pyright clients will be started by navigator which also means for example that every diagnostic is shown multiple times.
